--- a/layout/svg/base/src/nsSVGUtils.cpp	2012-07-11 03:33:17 +0800
+++ b/layout/svg/base/src/nsSVGUtils.cpp	2012-07-12 15:46:16 +0800
@@ -1203,6 +1203,19 @@
   if (opacity != 1.0f || maskFrame || (clipPathFrame && !isTrivialClip)) {
     complexEffects = true;
     gfx->Save();
+    if (!(aFrame->GetStateBits() & NS_STATE_SVG_NONDISPLAY_CHILD)) {
+      // aFrame has a valid visual overflow rect, so clip to it before calling
+      // PushGroup() to minimize the size of the surfaces we'll composite:
+      gfxContextMatrixAutoSaveRestore matrixAutoSaveRestore(gfx);
+      gfx->Multiply(GetCanvasTM(aFrame));
+      nsRect overflowRect = aFrame->GetVisualOverflowRectRelativeToSelf();
+      if (aFrame->IsFrameOfType(nsIFrame::eSVGGeometry)) {
+        // Unlike containers, leaf frames do not include GetPosition() in
+        // GetCanvasTM().
+        overflowRect = overflowRect + aFrame->GetPosition();
+      }
+      aContext->IntersectClip(overflowRect);
+    }
     gfx->PushGroup(gfxASurface::CONTENT_COLOR_ALPHA);
   }
 
