diff -r 1dc6cd15f683 configure.in
--- a/configure.in	Mon Feb 27 18:06:47 2012 -0600
+++ b/configure.in	Fri Mar 02 09:55:29 2012 +0900
@@ -9056,7 +9056,7 @@
     fi
     if test "$MOZ_OPTIMIZE" = "1"; then
         ac_configure_args="$ac_configure_args --enable-optimize"
-    else
+    elif test -z "$MOZ_OPTIMIZE"; then
         ac_configure_args="$ac_configure_args --disable-optimize"
     fi
     if test -n "$HAVE_64BIT_OS"; then
