diff --git a/config/config.mk b/config/config.mk
--- a/config/config.mk
+++ b/config/config.mk
@@ -608,7 +608,7 @@
 ifdef GNU_CC
 WIN32_EXE_LDFLAGS    += -mwindows
 else
-WIN32_EXE_LDFLAGS    += -SUBSYSTEM:WINDOWS
+WIN32_EXE_LDFLAGS    += -SUBSYSTEM:WINDOWS,5.01
 endif
 endif
 endif

diff --git a/configure.in b/configure.in
--- a/configure.in
+++ b/configure.in
@@ -2138,7 +2138,7 @@
         MKCSHLIB='$(LD) -NOLOGO -DLL -OUT:$@ -PDB:$(LINK_PDBFILE) $(DSO_LDOPTS)'
         MKSHLIB_FORCE_ALL=
         MKSHLIB_UNFORCE_ALL=
-        DSO_LDOPTS=-SUBSYSTEM:WINDOWS
+        DSO_LDOPTS=-SUBSYSTEM:WINDOWS,5.01
         _USE_CPP_INCLUDE_FLAG=1
         _DEFINES_CFLAGS='-FI $(DEPTH)/dist/include/mozilla-config.h -DMOZILLA_CLIENT'
         _DEFINES_CXXFLAGS='-FI $(DEPTH)/dist/include/mozilla-config.h -DMOZILLA_CLIENT'

diff --git a/js/src/config/config.mk b/js/src/config/config.mk
--- a/js/src/config/config.mk
+++ b/js/src/config/config.mk
@@ -608,7 +608,7 @@
 ifdef GNU_CC
 WIN32_EXE_LDFLAGS    += -mwindows
 else
-WIN32_EXE_LDFLAGS    += -SUBSYSTEM:WINDOWS
+WIN32_EXE_LDFLAGS    += -SUBSYSTEM:WINDOWS,5.01
 endif
 endif
 endif

diff --git a/js/src/configure.in b/js/src/configure.in
--- a/js/src/configure.in
+++ b/js/src/configure.in
@@ -1761,7 +1761,7 @@
         MKCSHLIB='$(LD) -NOLOGO -DLL -OUT:$@ -PDB:$(LINK_PDBFILE) $(DSO_LDOPTS)'
         MKSHLIB_FORCE_ALL=
         MKSHLIB_UNFORCE_ALL=
-        DSO_LDOPTS=-SUBSYSTEM:WINDOWS
+        DSO_LDOPTS=-SUBSYSTEM:WINDOWS,5.01
         _USE_CPP_INCLUDE_FLAG=1
         _DEFINES_CFLAGS='-FI $(DEPTH)/js-confdefs.h -DMOZILLA_CLIENT'
         _DEFINES_CXXFLAGS='-FI $(DEPTH)/js-confdefs.h -DMOZILLA_CLIENT'

diff --git a/security/coreconf/WIN32.mk b/security/coreconf/WIN32.mk
--- a/security/coreconf/WIN32.mk
+++ b/security/coreconf/WIN32.mk
@@ -145,7 +145,7 @@
 else # !NS_USE_GCC
     OS_CFLAGS += -W3 -nologo -D_CRT_SECURE_NO_WARNINGS \
          -D_CRT_NONSTDC_NO_WARNINGS
-    OS_DLLFLAGS += -nologo -DLL -SUBSYSTEM:WINDOWS
+    OS_DLLFLAGS += -nologo -DLL -SUBSYSTEM:WINDOWS,5.01
     ifeq ($(_MSC_VER),$(_MSC_VER_6))
     ifndef MOZ_DEBUG_SYMBOLS
     OS_DLLFLAGS += -PDB:NONE
 
diff --git a/nsprpub/configure.in b/nsprpub/configure.in
--- a/nsprpub/configure.in
+++ b/nsprpub/configure.in
@@ -1963,7 +1963,7 @@
             fi
         fi
 
-        OS_DLLFLAGS="-nologo -DLL -SUBSYSTEM:WINDOWS"
+        OS_DLLFLAGS="-nologo -DLL -SUBSYSTEM:WINDOWS,5.01"
         if test "$MSC_VER" -le "1200" -a -z "$MOZ_DEBUG_SYMBOLS"; then
             OS_DLLFLAGS="$OS_DLLFLAGS -PDB:NONE"
         fi

