diff -r ab1685df6e2f widget/src/windows/nsWindow.cpp
--- a/widget/src/windows/nsWindow.cpp    Sun Feb 05 14:48:18 2012 +0100
+++ b/widget/src/windows/nsWindow.cpp    Wed Feb 08 14:36:39 2012 +0900
@@ -7325,15 +7325,17 @@
     }
   }
 
-  // If a plugin is not visibile, especially if it is in a background tab,
+  // If a plugin is not visible, especially if it is in a background tab,
   // it should not be able to steal keyboard focus.  This code checks whether
   // the region that the plugin is being clipped to is NULLREGION.  If it is,
   // the plugin window gets disabled.
   if(mWindowType == eWindowType_plugin) {
     if(NULLREGION == ::CombineRgn(dest, dest, dest, RGN_OR)) {
+      ::ShowWindow(mWnd, SW_HIDE);
       ::EnableWindow(mWnd, FALSE);
     } else {
       ::EnableWindow(mWnd, TRUE);
+      ::ShowWindow(mWnd, SW_SHOW);
     }
   }
   if (!::SetWindowRgn(mWnd, dest, TRUE)) {

