diff -r 57eb38613ff8 widget/gtk2/nsGtkIMModule.cpp
--- a/widget/gtk2/nsGtkIMModule.cpp	Fri Feb 17 15:13:32 2012 -0800
+++ b/widget/gtk2/nsGtkIMModule.cpp	Tue Feb 21 15:33:14 2012 +0900
@@ -1492,6 +1492,35 @@
         return NS_ERROR_FAILURE;
     }
 
+    // Remove composition string
+    if (mIsComposing && !mDispatchedCompositionString.IsEmpty()) {
+        PRUint32 compLength = mDispatchedCompositionString.Length();
+        textContent.Cut(mCompositionStart, compLength);
+
+        // Adjust selLen
+        PRUint32 selEnd = selOffset + selLength;
+        PRUint32 compEnd = mCompositionStart + compLength;
+        PRUint32 selStartInComposition =
+          NS_MIN(NS_MAX(selOffset, mCompositionStart), compEnd);
+        PRUint32 selEndInComposition =
+          NS_MAX(NS_MIN(selEnd, compEnd), mCompositionStart);
+        if (selStartInComposition < selEndInComposition) {
+            selLength -= (selEndInComposition - selStartInComposition);
+        }
+
+        // Adjust selOffset
+        if (selOffset >= compEnd) {
+            selOffset -= compLength;
+        } else if (selOffset >= mCompositionStart) {
+            selOffset = mCompositionStart;
+        }
+        PR_LOG(gGtkIMLog, PR_LOG_ALWAYS,
+            ("    mCompositionStart=%u, compLength=%u, selOffset=%u, "
+             "selLength=%u, selStartInComposition=%u, selEndInComposition=%u",
+             mCompositionStart, compLength, selOffset, selLength,
+             selStartInComposition, selEndInComposition));
+    }
+
     // Get only the focused paragraph, by looking for newlines
     PRInt32 parStart = (selOffset == 0) ? 0 :
         textContent.RFind("\n", false, selOffset - 1, -1) + 1;
